From e8276be33f6a9437260b870cbe1dca81a4272bdd Mon Sep 17 00:00:00 2001
From: Tim Hartley <timothy.d.hartley@intel.com>
Date: Wed, 30 Oct 2013 13:20:08 -0700
Subject: Dalvik: Klocwork fix for dvmCompilerMIR2LIR error handler

BZ: 148776

The error handler returned by dvmCompilerArchSpecificNewCompilationErrorHandler
might be NULL, and will be dereferenced. This is a fatal error,
so this patch checks for the pointer == 0 and calls dvmCompilerAbort if it is.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I436aa1a1392210b96b48cd752a2d415a62f79578
Orig-MCG-Change-Id: I4bfff6aa5e823fd011b14855a1498e49d3665df0
Signed-off-by: Tim Hartley <timothy.d.hartley@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/BackEndEntry.cpp |    8 ++++++--
 vm/compiler/codegen/x86/X86Common.cpp    |    1 +
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/codegen/x86/BackEndEntry.cpp b/vm/compiler/codegen/x86/BackEndEntry.cpp
index e24bbdc..37e60ae 100644
--- a/vm/compiler/codegen/x86/BackEndEntry.cpp
+++ b/vm/compiler/codegen/x86/BackEndEntry.cpp
@@ -26,8 +26,12 @@ void dvmCompilerMIR2LIR (CompilationUnit *cUnit, JitTranslationInfo *info)
     //Create the error handler for the backend
     CompilationErrorHandler *errorHandler = dvmCompilerArchSpecificNewCompilationErrorHandler ();
 
-    //Paranoid
-    assert (errorHandler != 0);
+    if (errorHandler == 0)
+    {
+        // This is a fatal error to the compilation (and likely to the JIT)
+        ALOGE ("JIT: Could not create an error handler.");
+        dvmCompilerAbort (cUnit);
+    }
 
     //Set it in the cUnit
     cUnit->errorHandler = errorHandler;
diff --git a/vm/compiler/codegen/x86/X86Common.cpp b/vm/compiler/codegen/x86/X86Common.cpp
index 791b33c..68520cf 100644
--- a/vm/compiler/codegen/x86/X86Common.cpp
+++ b/vm/compiler/codegen/x86/X86Common.cpp
@@ -39,6 +39,7 @@ CompilationErrorHandler *dvmCompilerArchSpecificNewCompilationErrorHandler (void
         return fctPtr ();
     }
 
+    // TODO return some reasonable default error handler
     return 0;
 }
 
-- 
1.7.4.1

